### DISCUSSION TITLE
isinstance(..., dict) assumes unknown parameters

### DISCUSSION THREAD

**Discussion by AlanBogarin** [microsoft/pyright#9376] [Q&A]
I tried this feature with mypy, and there is no problem, because it assumes that the object is of type dict[Any, Any]
But Pyright assumes as dict[Unknown, Unknown]

```python
import json
import typing

data = json.loads("...")

if isinstance(data, dict):
    typing.reveal_type(data)  # Pyright: dict[Unknown, Unknown], Mypy: dict[Any, Any]
else:
    typing.reveal_type(data)
```
How can I check that the object is a dict without getting this error in strict mode?

**Comment by erictraut** [comment]
Mypy doesn't distinguish between explicit `Any` and implicit `Any`, which is what pyright calls `Unknown`.

The way I typically avoid generating an `Unknown` type with an `isinstance` conditional is to make sure that the pre-narrowed type (i.e. the type of the expression prior to the `isinstance` check) is a union of all expected types.

```python
# pyright: strict

type JSON = dict[str, JSON] | list[JSON] | str | float | None

data: JSON = json.loads("...")

if isinstance(data, dict):
    reveal_type(data)
```
